src/app/business/routes/algo/sagas/index.js

```diff
@@ -27,9 +30,35 @@ function* fetchList(request) {
         }
     }
 
-    if (!jwt) { // redirect to login page
-        yield put(actions.list.failure());
-        yield put(signOut.success());
+    if (!jwt) {
+        // try to refresh token
+        const {res, error} = yield call(fetchRefresh);
+
+        // refresh token does not exist
+        if (error) { // redirect to login page
+            yield put(refreshActions.failure(error));
+            yield put(actions.list.failure());
+            yield put(signOut.success());
+        }
+        else {
+            yield put(refreshActions.success(res));
+
+            if (typeof window !== 'undefined') {
+                const cookies = cookie.parse(window.document.cookie);
+                if (cookies['header.payload']) {
+                    jwt = cookies['header.payload'];
+                }
+            }
+
+            if (!jwt) {
+                yield put(actions.list.failure());
+                yield put(signOut.success());
+            }
+            else {
+                const f = () => fetchListApi(state.location.query, jwt);
+                yield call(fetchListSaga(actions, f), request);
+            }
+        }
```

Comment:
I have no doubt that this is working, but even within this function there is a bit of code duplication. What do you think of splitting it into a part that handles the jwt refresh and a part that only deals with the fetchList? Something like:

```js
function getJWTFromCookie() {
    let jwt;
    if (typeof window !== 'undefined') {
        const cookies = cookie.parse(window.document.cookie);
        if (cookies['header.payload']) {
            jwt = cookies['header.payload'];
        }
    }
    return jwt;
}


function* getJWT() {
    let jwt = getJWTFromCookie();


    if (!jwt) {
        const {res, error} = yield call(fetchRefresh);

        // refresh token does not exist
        if (error) { // redirect to login page
            yield put(refreshActions.failure(error));
            yield put(signOut.success());
        }
        else {
            yield put(refreshActions.success(res));
            jwt = getJWTFromCookie();
        }
    }

    return jwt;
}

function* fetchList(request) {
    const state = yield select();
    const jwt = yield getJWT();

    if (!jwt) {
        yield put(actions.list.failure());
    }
    else {
        const f = () => fetchListApi(state.location.query, jwt);
        yield call(fetchListSaga(actions, f), request);
    }
}
```

Reply:
We can refacto some of this, but actions need to be in a specific order.
`signout` should be after `list.failure`